src/proj.rs

```diff
@@ -196,15 +199,17 @@ pub trait Info {
     /// # Safety
     /// This method contains unsafe code.
     fn info(&self) -> Result<Projinfo, ProjError> {
-        let pinfo: PJ_INFO = unsafe { proj_info() };
-        Ok(Projinfo {
-            major: pinfo.major,
-            minor: pinfo.minor,
-            patch: pinfo.patch,
-            release: _string(pinfo.release)?,
-            version: _string(pinfo.version)?,
-            searchpath: _string(pinfo.searchpath)?,
-        })
+        unsafe {
```

Comment:
What I've gathered from your other changes in this PR is that you advocate marking the function as `unsafe fn` when the caller of the function needs to ensure some invariant is upheld, whereas if all unsafe invariants are intended to be handled within the function, so the caller needn't consider them, you don't mark the function as `unsafe fn`.

Is that right? I've never really considered it, but that makes sense to me.

Reply:
> you advocate marking the function as `unsafe fn` when the caller of the function needs to ensure some invariant is upheld, whereas if all unsafe invariants are intended to be handled within the function, so the caller needn't consider them, you don't mark the function as `unsafe fn`

Yeah exactly. And I think this section of TRPL covers that better than I can explain:

https://doc.rust-lang.org/nightly/book/ch19-01-unsafe-rust.html#creating-a-safe-abstraction-over-unsafe-code

Reply:
I'm permanently confused about when / when not to mark functions unsafe.

Reply:
When determining whether a function should be marked `unsafe`, I ask myself: when a user calls this function, is it possible for the user to provide inputs to the function such that it could result in any of [the behaviors Rust considers undefined](https://doc.rust-lang.org/nightly/nomicon/what-unsafe-does.html), which includes:

> Dereferencing (using the `*` operator on) a dangling or unaligned raw pointer.

and

> ...raw pointer read from uninitialized memory, ...

So consider this function (taken directly from [the FFI page in the Nomicon](https://doc.rust-lang.org/nightly/nomicon/ffi.html)):

```rust
fn kaboom(ptr: *const i32) -> i32 {
    *ptr
}
```

I'm able to call this function with whatever pointer I want and it will attempt to read from that memory location (e.g. something like: `kaboom(0x1a1a as *const i32)` will attempt to read a `i32` at memory address `0x1a1a`). So I am able to call this function with a certain input that will cause it to perform undefined behavior (in this case read from uninitialized memory). Thus we need to mark the function as `unsafe`.

Coming back to PROJ-land, consider this example:

```rust
/// Print the description of a CRS given a definition string
fn crs_description(crs_definition: CString) {
    let pj_ptr = proj_sys::proj_create(0, crs_definition.as_ptr());
    let pj_proj_info = proj_sys::proj_pj_info(pj_ptr);
    println!("Description: {}", _string(pj_proj_info.description));
}
```

Is it possible to call this function and have it succeed? Yes! Pass it a valid `crs_definition` that we know has a description and it should print.

Is it possible to call this function that will result in undefined behavior? Yes! Pass it a `crs_definition` PROJ doesn't know about and `proj_create` will return a null pointer, and then `proj_sy::proj_pj_info` will attempt to deference the null pointer, which is considered undefined. Thus we would need to mark this function as unsafe. And since it's marked as `unsafe` we would add a section in the docs for this function `# Safety` that describes how you could use this function in a safe way (only passing it valid CRS definitions).

And one more example:

```rust
/// Print the description of a CRS given a definition string
fn crs_description(crs_definition: CString) {
    let pj_ptr = proj_sys::proj_create(0, crs_definition.as_ptr());
    if pj_ptr.is_null() {
        return;
    }
    let pj_proj_info = proj_sys::proj_pj_info(pj_ptr);
    if pj_proj_info.description.is_null() {
        return;
    }
    println!("Description: {}", _string(pj_proj_info.description));
}
```

Is it possible to call this function that will result in undefined behavior? Assuming I'm not overlooking anything, I don't think so! You can pass it whatever string you want and we shouldn't encounter any undefined behavior. Thus we _don't_ need to mark it as unsafe.

Hope that helps a little!